lib/metrics.js

```diff
@@ -0,0 +1,129 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const EXPORTED_SYMBOLS = ['Metrics'];
+
+function Metrics(opts) {
+  /*
+  The metrics class is used to determine whether the recommendation is used
+  when the user navigates with the popup open. Data is sent to the Test Pilot
+  add-on, which forwards it to the Telemetry server.
+  */
+  this.events = opts.events;
+  this.interaction = null;
+  this.observerService = opts.observerService;
+  this.win = opts.win;
+  this.recommendation = opts.recommendation;
+  this.popup = opts.popup;
+  this.checked = false;
+
+  this.interactionStart = this.interactionStart.bind(this);
+  this.interactionEnd = this.interactionEnd.bind(this);
+  this.onKeyNavigate = this.onKeyNavigate.bind(this);
+  this.onClickNavigate = this.onClickNavigate.bind(this);
+}
+
+Metrics.prototype = {
+  init: function() {
+    this.win.console.log('metrics.init');
+    // Start listening when the popup is about to open.
+    this.events.subscribe('before-popup-show', this.interactionStart);
+
+    // Detect when a navigation event occurs.
+    this.events.subscribe('before-keyboard-navigate', this.onKeyNavigate);
+    this.events.subscribe('before-click-navigate', this.onClickNavigate);
+
+    // Reset the interaction state when the urlbar contents change.
+    this.events.subscribe('urlbar-change', this.interactionStart);
+
+    // The interaction ends when the popup closes.
+    this.events.subscribe('before-popup-hide', this.interactionEnd);
+  },
+  destroy: function() {
+    this.win.console.log('metrics.destroy');
+    this.events.unsubscribe('before-popup-show', this.interactionStart);
+    this.events.unsubscribe('before-keyboard-navigate', this.onKeyNavigate);
+    this.events.unsubscribe('before-click-navigate', this.onClickNavigate);
+    this.events.unsubscribe('urlbar-change', this.interactionStart);
+    this.events.unsubscribe('before-popup-hide', this.interactionEnd);
+  },
+  interactionStart: function() {
+    this.win.console.log('metrics.interactionStart');
+    // Initialize default values for the case where the popup closes without a
+    // navigation.
+    this.checked = false;
+    this.interaction = {
+
+      // Did the user navigate?
+      didNavigate: false,
+
+      // Was the selected item clicked or keyed or neither?
+      interactionType: null,
+
+      // Was the recommendation shown?
+      recommendationShown: false,
+
+      // What type of recommendation was it?
+      recommendationType: null,
+
+      // Was the recommendation the selected item?
+      recommendationSelected: null,
+
+      // What was the index of the selected item?
+      // If the recommendation was selected, this will be -1.
+      selectedIndex: null
```

Comment:
@nchapman Does this look like it answers the basic product questions? This packet is constructed each time the popup opens, and is fired each time the popup closes, regardless of whether the user navigated or interacted with the popup while it was open. One missing bit of data that I'll add is the wikipedia vs TLD type.

This data should be enough for us to calculate the click-through rate of the popup with and without the recommendation visible, and drill down into click-through rates of the different types of recommendations.

We should also be able to figure out how often users interact with the popup, since it always pops open when you enter stuff in the urlbar.